### PR TITLE
Add "None of these" and "or" divider to checkboxes

### PR DESCRIPTION
**Pull request by frankieroberto**
Following [conversation about the checkboxes component](https://github.com/alphagov/govuk-design-system-backlog/issues/37), this adds the "divider" feature from the Radios component to the Checkboxes component.

The main use for the divider is to add an "or" before a "None of these option", which allows a user to explicitly state that none of the checkboxes apply, rather than having to infer (or be given guidance) that they can leave all the checkboxes unticked.  This has the additional benefit of being able to validate that the user has selected at least one checkbox, avoiding a potential issue where a user accidentally doesn't check a relevant checkbox, and the service assuming this meant that none of them were relevant. Services should also check that the "None of these" option wasn't ticked as well as another checkbox, and if so display a relevant error message.

In order to help users avoid accidentally selecting an option as well as "None of these", some javascript is included which prevents this from happening, by unchecking all other checkboxes when checking a "None of these" checkbox, and by unchecking the "None of these" checkbox when checking any other checkbox. This behaviour can be enabled by passing `behaviour: "exclusive"` as a param on a checkbox item.

➡️ [PR which adds guidance on this to the Design System website](https://github.com/alphagov/govuk-design-system/pull/1535)

## Example showing use within Nunjucks macro:

```nunjucks
{{ govukCheckboxes({
  idPrefix: "waste",
  name: "waste",
  fieldset: {
    legend: {
      text: "Which types of waste do you transport?",
      isPageHeading: true,
      classes: "govuk-fieldset__legend--l"
    }
  },
  hint: {
    text: "Select all that apply."
  },
  items: [
    {
      value: "carcasses",
      text: "Waste from animal carcasses"
    },
    {
      value: "mines",
      text: "Waste from mines or quarries"
    },
    {
      value: "farm",
      text: "Farm or agricultural waste"
    },
    {
      divider: "Or"
    },
    {
      value: "none",
      text: "None of these",
      behaviour: "exclusive"
    }
  ]
}) }}
```

## Example HTML output

```html
<div class="govuk-form-group">
<fieldset class="govuk-fieldset">
  <legend class="govuk-fieldset__legend govuk-fieldset__legend--l">
    <h1 class="govuk-fieldset__heading">
      Which types of waste do you transport regularly?
    </h1>
  </legend>
  <div class="govuk-checkboxes" data-module="govuk-checkboxes">
          <div class="govuk-checkboxes__item">
            <input class="govuk-checkboxes__input" id="with-divider-and-none" name="with-divider-and-none" type="checkbox" value="animal">
            <label class="govuk-label govuk-checkboxes__label" for="with-divider-and-none">
        Waste from animal carcasses
      </label>
          </div>
          <div class="govuk-checkboxes__item">
            <input class="govuk-checkboxes__input" id="with-divider-and-none-2" name="with-divider-and-none" type="checkbox" value="mines">
            <label class="govuk-label govuk-checkboxes__label" for="with-divider-and-none-2">
        Waste from mines or quarries
      </label>
          </div>
          <div class="govuk-checkboxes__item">
            <input class="govuk-checkboxes__input" id="with-divider-and-none-3" name="with-divider-and-none" type="checkbox" value="farm">
            <label class="govuk-label govuk-checkboxes__label" for="with-divider-and-none-3">
        Farm or agricultural waste
      </label>
          </div>
          <div class="govuk-checkboxes__divider">or</div>
          <div class="govuk-checkboxes__item">
            <input class="govuk-checkboxes__input" id="with-divider-and-none-5" name="with-divider-and-none" type="checkbox" value="none" data-behaviour="exclusive">
            <label class="govuk-label govuk-checkboxes__label" for="with-divider-and-none-5">
        None of these
      </label>
          </div>
  </div>

</fieldset>
</div>
```

# Examples of the checkbox divider in use on live services

79326612-0a284300-7f0b-11ea-82e3-df502e7d052b

78349670-dc9ed980-759b-11ea-9175-9cd61f196839

Screenshot 2021-04-08 at 14 00 27

### CODE REVIEW

**Comment by edwardhorsford** [comment]
Very pleased to see this. I've been tripped up several times confused why my separator wasn't working.

**Comment by timpaul** [comment]
Following a conversation with the team and with @frankieroberto - we've agreed that Frankie will add a JavaScript enhancement to the PR, that _unchecks_ all the other checkboxes, if a 'None of the above' checkbox is checked.

Before @frankieroberto starts work though, we'll catch up to agree on the Nunjucks macro API - the card for that is here: https://github.com/alphagov/govuk-frontend/issues/2152

**Comment by frankieroberto** [comment]
This is no longer blocked, now that we’ve agreed the API in #2152.

I've done the initial javascript implementation. Would welcome any initial feedback or suggestions. I tried to avoid making too many unrelated changes to the checkboxes javascript (eg to align it more closely with the radios javascript), but that could be considered.

I’ll write the tests next.

**Comment by frankieroberto** [comment]
I've now added some tests, both to test the macro template output, and to test the checking/unchecking behaviour.

**Comment by calvin-lau-sig7** [comment]
[Added suggestions to guidance content in the PR](https://github.com/alphagov/govuk-design-system/pull/1535) ahead of going to review by the Design System working group.

**Comment by frankieroberto** [comment]
@calvin-lau-sig7 thanks! I've made some changes based on your suggestions. Have also added a paragraph about labelling the option.

**Comment by adamsilver** [comment]
I recognise the need for users to disclose ‘none of the above’ but I’m concerned with the specific solution.

Taking this question as an example:

```
Which of these qualifications do you have?
[x] GCSE
[x] Degree
[ ] A Levels
[x] None of the above

[Continue]
```

- ‘None of the above’ is not a type of qualification. Because of this I think it'll increase cognitive load and users with cognitive impairments will especially struggle
- It looks like you can pick ‘None of the above’ as well as other options. And you incorrectly can without JavaScript - which leads to funny error messages
- With JavaScript I presume it will automatically uncheck options but this is not how checkboxes work conventionally. Perhaps this will be unexpected by certain users. I wonder if this is inaccessible to some degree.

My other concern with this is that by putting it in the design system we're encouraging the use of this pattern over other, potentially clearer/more accessible solutions. See 3 alternative examples below.

## Example 1: adding a guard question beforehand

```
Do you have any qualifications?
( ) Yes
( ) No

[Continue]
```

## Example 2: Marking/hinting the question as optional

Here I’ve used hint text but you could put something in the legend so it's harder to miss.

```
Which of these qualifications do you have?
Select continue if you have none of these options
[ ] Option 1
[ ] Option 2
[ ] Option 3

[Continue]
```

## Example 3: Having an explicit option not as a checkbox

This could be more useful when the user needs to specifically disclose ‘none of the above’.

```
Which of these qualifications do you have?
[ ] Option 1
[ ] Option 2
[ ] Option 3

[Continue]
I don't have any of these qualifications
```

**Comment by StephenGill** [comment]
Just as a variation on @adamsilver's example #3 - in line with the new [equality information pattern](https://design-system.service.gov.uk/patterns/equality-information/), you could say something like 'If you do not have any of these qualifications, continue without checking any of the boxes'.

```
What is your date of birth?

For example, 31 3 1980. If you prefer not to say, continue without entering any information.

Day     Month       Year
[       ].  [         ]    [        ]

[Continue]

```

**Comment by joelanman** [comment]
Hey @adamsilver!

>‘None of the above’ is not a type of qualification.

No, but it is a reasonable answer to the question "Which of these qualifications do you have?" isn't it?

>It looks like you can pick ‘None of the above’ as well as other options.

I think the 'or' divider which you're missing from your example helps here. Plus I'm not sure anyone would expect to be able to select None and another option - that wouldn't make sense, so I don't think it's relying on the 'or' only.

On your alternatives:

1.  Sometimes a guard question would be better, I agree. However in some cases a guard question cannot be clear - and this is a good example. We're not asking if you have any qualifications, we're asking if you have any in a specific list.

2. The problem with this approach is you can't do any validation - someone could accidentally continue without actually answering the question - this could cause real issues for the user.

3. I think this is probably less accessible? The user would consider the options presented in the checkboxes and may not realise there is a separate option available.

**Comment by adamsilver** [comment]
Hey @joelanman - all good points...

> No, but it is a reasonable answer to the question "Which of these qualifications do you have?" isn't it?

I appreciate this is not a clear cut thing. Perhaps as a slight variant:

```
Do you have any of the following qualifications?
- Qualification 1
- Qualification 2
- Qualification 3
( ) Yes, I have one or more of these
( ) No, I do not have any of these particular qualifications

[Continue]
```

> I think the 'or' divider which you're missing from your example helps here. Plus I'm not sure anyone would expect to be able to select None and another option - that wouldn't make sense, so I don't think it's relying on the 'or' only.

I’m not sure, but I think the checkbox signifier (that you can select multiple) shouldn't be messed with ideally. I also did make an edit to my response to acknowledge the value of the divider to _sighted_ users.

> 1. Sometimes a guard question would be better, I agree. However in some cases a guard question cannot be clear - and this is a good example. We're not asking if you have any qualifications, we're asking if you have any in a specific list.

Yep - the reason I mention it is because at the very least I think the design system needs to state other options before this one (if this solution does end up being added)

> 2. The problem with this approach is you can't do any validation - someone could accidentally continue without actually answering the question - this could cause real issues for the user.

Yep - but I would say that's the same for any field that's optional. Perhaps the ‘if you have any’ could be added to the legend.

> 3. I think this is probably less accessible? The user would consider the options presented in the checkboxes and may not realise there is a separate option available.

Yep, I do not love this one. But then I haven't tested/iterated such a pattern in research.

I guess I just feel that something like this should be tested more thoroughly with a range of users with different abilities before being added to the design system. And perhaps alongside a range of approaches that might be better depending on the context.

**Comment by joelanman** [comment]
@adamsilver good points!

Agree we should highlight the alternatives, like we do on other patterns like accordions and tabs for example - if a guard question can be simple its probably the better option (Do you have a car > What kind of car do you have)

> Yep - but maybe that's the same for any field that's optional

I think we need to make the distinction between a required question where the answer can be 'None of these' (we need the answer) and an optional question where we don't need an answer. So a question with 'None' could easily be a required question.

**Comment by joelanman** [comment]
worth noting that we used None of these on a Coronavirus service recently and it tested well, and it was used a lot on the recent census.

**Comment by edwardhorsford** [comment]
It would be good for the guidance to note that sometimes a guard question is a better solution - but as @joelanman notes, there are also lots of times where it's incompatible with what you're asking. 

I don't agree it increases cognitive load - if anything it's easier as you're making sure there's always an answer that the user can select from. If you were doing this data collection in conversation with a user / over the phone, it's the sort of answer you'd expect from them verbally.

I've seen this used in multiple services over the years (as well as ones I've worked on) - IMHO it's rather well tested.

**Comment by adamsilver** [comment]
Another idea would be to split up into separate questions like this:

```
Do you have a GCSE?
( ) Yes ( ) No

Do you have an A-level?
( ) Yes ( ) No

Do you have a degree?
( ) Yes ( ) No

[Continue]
```

**Comment by edwardhorsford** [comment]
Something else in favour of this pattern - this matches how paper forms often work. People are used to answering this sort of question.

From the UK Census 2021:
Screenshot 2021-04-14 at 12 00 33

**Comment by simonwhatley** [comment]
> Something else in favour of this pattern - this matches how paper forms often work. People are used to answering this sort of question.
> 
> From the UK Census 2021:
> Screenshot 2021-04-14 at 12 00 33

For me, the paper form looks like you can tick, for example, 'Working as an employee' and 'OR none of the above'.

If a user does select these items, this would require the Census team to interpret the user's answer or follow up with them.

In any case, it seems prudent to avoid this design.

**Comment by joelanman** [comment]
>For me, the paper form looks like you can tick, for example, 'Working as an employee' and 'OR none of the above'.

In the digital version proposed this would not be possible

**Comment by simonwhatley** [comment]
To digress from the early conversation around the need for this feature, I'd like to explore the choice of `behaviour: "exclusive"`. 

Exclusive suggests there's a `behaviour: "inclusive"` version yet to be implemented. If there isn't, a boolean option makes sense (`exclusive: true`, which is used in the GOV.UK Publishing Components equivalent). 

A boolean option maps more closely to the `checked`, `conditional` and `disabled` parameters used in the checkbox options, and the `open` and `isPageHeading` parameters used elsewhere.

**Comment by joelanman** [comment]
the idea is `behaviour` could potentially be extended to something like `Select all` functionality if we wanted to add that in the future.

**Comment by simonwhatley** [comment]
> > For me, the paper form looks like you can tick, for example, 'Working as an employee' and 'OR none of the above'.
> > If a user does select these items
> 
> In the digital version proposed this would not be possible

Yep, assuming JavaScript was available.

**Comment by simonwhatley** [comment]
> the idea is `behaviour` could potentially be extended to something like `Select all` functionality if we wanted to add that in the future.

Yep so "Select all" would be the 'inclusive' behaviour.

Checked, conditional, disabled and so on are also types of behaviour, yet given their own parameters. I'm wondering why 'exclusive' is being treated differently.

**Comment by joelanman** [comment]
you'd also validate on the server side

**Comment by 36degrees** [comment]
> Exclusive suggests there's a `behaviour: "inclusive"` version yet to be implemented. If there isn't, a boolean option makes sense (`exclusive: true`, which is used in the GOV.UK Publishing Components equivalent).

This has been discussed previously in https://github.com/alphagov/govuk-frontend/issues/2152#issuecomment-808354055:

> We'd prefer to avoid a boolean flag for a behaviour like this, as it can lead to strange 'undefined' functionality if we introduced other competing behaviours in the future, for example if we ended up with selectAll and selectNone booleans, users could technically pass true for both.

**Comment by simonwhatley** [comment]
> We'd prefer to avoid a boolean flag for a behaviour like this, as it can lead to strange 'undefined' functionality if we introduced other competing behaviours in the future, for example if we ended up with selectAll and selectNone booleans, users could technically pass true for both.

@36degrees If someone set competing booleans would you not ignore them both, or is that a bad component design? 

I also wonder whether there's a tradeoff between managing problems like this _under the hood_ and having a consistent Nunjucks interface where specific _features_ of a component are switched on/off with a boolean.

**Comment by adamsilver** [comment]
@36degrees if you had competing booleans, you could throw an error at build/run time - what do you think?

**Comment by joelanman** [comment]
```
Do you have a GCSE?
( ) Yes ( ) No

Do you have an A-level?
( ) Yes ( ) No

Do you have a degree?
( ) Yes ( ) No
```

I think using radios like this can work for short lists, and we could mention it in guidance. But my concern with longer lists would be a lot of repetition ('Do you have a') and it makes the user do more work - in the case of 'None' they have to click every no in a long list instead of a single "None of these".

**Comment by frankieroberto** [comment]
Hi all. Thanks for the comments (I was away last week).

I'd agree that the pattern/component may not be the best solution in all contexts, but I think we have enough evidence that there are some contexts where it helpful, namely where there is a list of things the user has to check before answering. The alternative is to allow users to leave all checkboxes unticked.

The advantages of having an explicit "No, I have none of these" answer are:

* Removes the need to have to tell people that they can press continue without selecting any in hint text (which is awkward to phrase, and increases cognitive load, especially as services largely avoid optional questions)
* Reduces errors caused by people clicking `[Continue]` having not selected anything by mistake (this might be rare, but could lead to significant issues for the user depending on the service)
* In some contexts, some users may _feel_ more comfortable with explicitly saying "none of these" rather than implicitly saying it by leaving all the checkboxes unticked – a user researcher in the Pension Credit [added the option for this reason](https://github.com/alphagov/govuk-design-system-backlog/issues/37#issuecomment-815872939)

There are some contexts where a guard question may be better though - eg if it avoids the user from having to check the list completely – and some where leaving them all unchecked might be fine (perhaps where there are regular or internal users).

The [draft guidance](https://github.com/alphagov/govuk-design-system/pull/1535) hopefully makes it clear that the wording of the "none" checkbox label should read as a possible answer to the question legend, should use "these" rather than "the above", and should be specific to the question rather than generic. I'd welcome any comments on how the guidance could be further improved though.

Regarding this point:

> With JavaScript I presume it will automatically uncheck options but this is not how checkboxes work conventionally. Perhaps this will be unexpected by certain users. I wonder if this is inaccessible to some degree.
– @adamsilver 

The [ONS implementation](https://ons-design-system.netlify.app/components/mutually-exclusive/) adds a visually hidden "Selecting this will uncheck all other checkboxes" to the label IF there are other checkboxes checked. It also moves the "Or" to the label for screen reader users. Worth exploring (and researching).

**Comment by frankieroberto** [comment]
> @36degrees if you had competing booleans, you could throw an error at build/run time - what do you think?
> – @adamsilver 

Currently I don't think any of the nunjucks macros ever throw any errors (even if required params are missing, or unknown params are added). I think this is by design – perhaps the exceptions are harder for less experienced users to debug?

**Comment by CharlotteDowns** [comment]
[Department for Work and Pensions have shared some initial findings](https://github.com/alphagov/govuk-design-system-backlog/issues/37#issuecomment-815872939) from User Research on a checkbox with a divider. For more information, I'm sure they would be happy for us to reach out.

**Comment by timpaul** [comment]
Review on 7 June

File: src/govuk/components/checkboxes/checkboxes.js
Comment (36degrees) [reply src/govuk/components/checkboxes/checkboxes.js:120]: This is also a very long line, so would be good to split it up.

```suggestion
  var allInputsWithSameNameAndExclusiveBehaviour = document.querySelectorAll(
    'input[data-behaviour="exclusive"][type="checkbox"][name="' + $input.name + '"]'
  )
```

**Review by 36degrees — COMMENTED** [review]
Generally this is looking really good – thanks @frankieroberto 🙌🏻 

Tested in:

- Internet Explorer 8
- Internet Explorer 9
- Internet Explorer 10
- Internet Explorer 11
- Chrome 91 (macOS)
- Edge 91 (Windows)
- Safari 14.1 (macOS)
- Firefox 89 (macOS)
- Safari (iOS 14.6)
- Samsung Browser 11.1 (Android 11)

Just a few minor comments from me. Once those have been addressed, this'll need a second review from another developer on the team before we can merge it.

File: src/govuk/components/checkboxes/template.njk
Comment (36degrees) [reply src/govuk/components/checkboxes/template.njk:30]: I am wondering if we could just simplify things by always including the `module` attribute – but if we were going to do that we should do it consistently, so this seems like something we should revisit in the future instead.

Comment (frankieroberto) [reply src/govuk/components/checkboxes/template.njk:30]: Good shout to combine them. I'll do that.

I'm guessing that not initialising the javascript when it's not needed saves a tiny bit of performance – but it might be fairly trivial?

Comment (frankieroberto) [reply src/govuk/components/checkboxes/template.njk]: Done in c401d6cc0770657784f5a21203b2d04991080943

File: src/govuk/components/checkboxes/checkboxes.js
Comment (adamsilver) [inline src/govuk/components/checkboxes/checkboxes.js:96]: @frankieroberto @36degrees I think this should be scoped by the module container. 

If this is left as is, if there were two different forms with the same name checkboxes (which is legit), things will break.

Comment (frankieroberto) [reply src/govuk/components/checkboxes/checkboxes.js:96]: @adamsilver I think it should probably be form-scoped instead, as you could use two calls to the checkboxes macro to build a single list of checkboxes? (Not sure why you would mind).

Alternatively, perhaps we shouldn't rely on the name attribute at all, as it would also be totally valid to give each checkbox a different `name` attribute value, but still have them be related to each other? In which case, maybe it makes sense to scope to the module and just search for all checkboxes within that? 🤔

Comment (frankieroberto) [reply src/govuk/components/checkboxes/checkboxes.js:96]: Worth noting that the conditional reveals currently [use module-scope](https://github.com/alphagov/govuk-frontend/blob/main/src/govuk/components/checkboxes/checkboxes.js#L79) to find the conditional content (by ID).

Comment (36degrees) [reply src/govuk/components/checkboxes/checkboxes.js:96]: > If this is left as is, if there were two different forms with the same name checkboxes (which is legit), things will break.

That's exactly what the `hasSameFormOwner` check is for. Effectively, we're trying to mimic the [algorithm defined in the HTML spec to determine if two radio buttons are in the same radio button group](https://html.spec.whatwg.org/multipage/input.html#radio-button-group).

> I think it should probably be form-scoped instead, as you could use two calls to the checkboxes macro to build a single list of checkboxes? (Not sure why you would mind).

The use case that's come up previously is if you want to group checkboxes / radios within subheadings, like in https://github.com/alphagov/govuk-frontend/issues/1079. At the minute the only way to do that whilst using the macros is to call it several times with the same name, which would result in multiple 'modules'.

> Worth noting that the conditional reveals currently [use module-scope](https://github.com/alphagov/govuk-frontend/blob/main/src/govuk/components/checkboxes/checkboxes.js#L79) to find the conditional content (by ID).

The conditional reveal 'has to' be in the same module, but other checkboxes with the same name do not.

Comment (frankieroberto) [reply src/govuk/components/checkboxes/checkboxes.js:96]: > That's exactly what the hasSameFormOwner check is for.

Opps, forgot about this, even though I wrote it. 🤦🏻‍♂️

File: src/govuk/components/checkboxes/template.njk
Comment (36degrees) [reply src/govuk/components/checkboxes/template.njk]: > I'm guessing that not initialising the javascript when it's not needed saves a tiny bit of performance – but it might be fairly trivial?

Prompted by @adamsilver's comment I just tested having multiple modules on the page and discovered a bit of an edge case which I think means we need to always include the module attribute.

If you have multiple groups of checkboxes with the same name on the page, but only some of those groups have the exclusive behaviour or a conditional reveal, checking one of the checkboxes in a groups without the module attribute on _does not_ uncheck the exclusive checkbox with the same name, because there are no event listeners set up on the group of checkboxes you checked.

For example, try checking "None of the above" in this example, then checking one of the colours in the other groups.

```jinja
{% call govukFieldset({
    legend: {
    text: "Which colours do you like?",
    classes: "govuk-fieldset__legend--l"
    }
})%}

  <h2 class="govuk-heading-m">Primary colours</h2>

  {{ govukCheckboxes({
    idPrefix: "colour-primary",
    name: "colour",
    items: [
      {
        value: "red",
        text: "Red"
      },
      {
        value: "yellow",
        text: "Yellow"
      },
      {
        value: "blue",
        text: "Blue"
      }
    ]
  }) }}

  <h2 class="govuk-heading-m">Secondary colours</h2>

  {{ govukCheckboxes({
    idPrefix: "colour-secondary",
    name: "colour",
    items: [
      {
        value: "green",
        text: "Green"
      },
      {
        value: "purple",
        text: "Purple"
      },
      {
        value: "orange",
        text: "Orange"
      }
    ]
  }) }}

  <h2 class="govuk-heading-m">Other colours</h2>

  {{ govukCheckboxes({
    idPrefix: "colour-other",
    name: "colour",
    items: [
      {
        value: "imaginary",
        text: "An imaginary colour"
      },
      {
        divider: "or"
      },
      {
        value: "none",
        text: "None of the above",
        behaviour: "exclusive"
      }
    ]
  }) }}

{% endcall %}
```

Comment (frankieroberto) [reply src/govuk/components/checkboxes/template.njk]: @36degrees nice spot. Shall we remove the logic and always include the `data-module="govuk-radio"` attribute?

Comment (frankieroberto) [reply src/govuk/components/checkboxes/template.njk]: @36degrees I've fixed this issue by always including the `data-module` attribute here: 0629a0a571baa129bae07ed9b43baa21fe3d97f5

**Comment by 36degrees** [comment]
This is looking good @frankieroberto, thanks.

It'd be good to tidy up the commit history a bit, as I think quite a few of the commits could be squashed without losing anything important, and there's a few merge commits in there that ideally we'd lost. However, I think it's worth getting a second review from another member of the team first, in case anything else comes up.

Does that sound sensible?

**Comment by frankieroberto** [comment]
@36degrees second review sounds good. 👍 

Happy to force push some squashed commits at the end (I tend to avoid doing that as it can break things for anyone else who's fetched the branch), but every project has its own policy!

**Review by lfdebrux — APPROVED** [review]
I've given this a quick once-over: looked at the code; tested in Chrome, Firefox and Safari on macOS, and Chrome on Android (real device).

I couldn't find any issues, it all looks good to me!

**Comment by 36degrees** [comment]
Brill, thanks @lfdebrux for reviewing 🙌🏻 

@frankieroberto are you able to look at tidying up the commit structure now?

**Comment by frankieroberto** [comment]
@36degrees squashed into 5 commits (and reordered them a bit). Does that look ok? Been a while since I did so much command-line git... 😄

**Comment by 36degrees** [comment]
That's great, thanks for sorting that @frankieroberto 👍🏻

**Comment by edwardhorsford** [comment]
@a184studio from a different perspective, with javascript we can _avoid_ an error message. Which should hopefully make things simpler for users.

**Comment by a184studio** [comment]
Has anyone done anything on checkboxes being a positive action vs a negative one. 

I'm investigating the ability to check items that are wrong vs right. 
The initial analysis that we have done suggested that the minority will be right. 

Which in reality means that they would be selecting 24/25 checkboxes. [Tick] 
Or you do it the other way, Select all that are wrong. Which you would have to interact 1/25 times. [Tick, but given ticks as an icon = good]

**EG.** 

Select all that are incorrect 
[ Tick ] // (Which is which is jarring as a tick = positive, but the intention is negative outcome. This way allows us to measure the error, which will be the minority)
**vs**
Select all that are correct. 
[ Tick ] (to measure the number of things incorrect, you would have to measure the volumes of none selected items)

Any insights are grateful.

**Comment by 36degrees** [comment]
@a184studio this PR has now been closed, and the only people that are likely to see your comment are the people who have previously contributed to this discussion.

Can I suggest instead adding a comment to the [backlog discussion for checkboxes](https://github.com/alphagov/govuk-design-system-backlog/issues/37), where more people are likely to see it?

For more visibility, it might also be worth asking in the [#design channel on cross-government Slack](https://ukgovernmentdigital.slack.com/archives/C062GAGLW), if that is something you have access to.

**Pull request merged**
